backend/controller/admin/admin_test.go

```diff
@@ -0,0 +1,151 @@
+	ctx := log.ContextWithNewDefaultLogger(context.Background())
+
+	cm, err := cf.New(ctx,
+		cf.ProjectConfigResolver[cf.Configuration]{Config: []string{config}},
+		[]cf.Provider[cf.Configuration]{
+			cf.EnvarProvider[cf.Configuration]{},
+			cf.InlineProvider[cf.Configuration]{},
+		})
+	assert.NoError(t, err)
+
+	// Skip for now, this is failing in GH CI and I give up.
+	// kcp := cf.KeychainProvider{}
+	// _, err = kcp.Store(ctx, cf.Ref{Name: "mutable"}, []byte("\"helloworldsecret\""))
+	// assert.NoError(t, err)
+
+	sm, err := cf.New(ctx,
+		cf.ProjectConfigResolver[cf.Secrets]{Config: []string{config}},
+		[]cf.Provider[cf.Secrets]{
+			cf.EnvarProvider[cf.Secrets]{},
+			cf.InlineProvider[cf.Secrets]{},
+			// kcp,
+		})
+	assert.NoError(t, err)
+	admin := NewAdminService(cm, sm)
+	assert.NotZero(t, admin)
+
+	expectedEnvarValue, err := json.MarshalIndent(map[string]string{"bar": "barfoo"}, "", "  ")
+	assert.NoError(t, err)
+
+	testAdminConfigs(t, ctx, "FTL_CONFIG_YmFy", admin, []expectedEntry{
+		{Ref: cf.Ref{Name: "bar"}, Value: string(expectedEnvarValue)},
+		{Ref: cf.Ref{Name: "foo"}, Value: "\"foobar\""},
+		{Ref: cf.Ref{Name: "mutable"}, Value: "\"helloworld\""},
+		{Ref: cf.Ref{Module: optional.Some[string]("echo"), Name: "default"}, Value: "\"anonymous\""},
+	})
+
+	testAdminSecrets(t, ctx, "FTL_SECRET_YmFy", admin, []expectedEntry{
+		{Ref: cf.Ref{Name: "bar"}, Value: string(expectedEnvarValue)},
+		{Ref: cf.Ref{Name: "foo"}, Value: "\"foobarsecret\""},
+		// {Ref: cf.Ref{Name: "mutable"}, Value: "\"helloworldsecret\""},
+	})
+}
+
+type expectedEntry struct {
+	Ref   cf.Ref
+	Value string
+}
+
+func tempConfigPath(t *testing.T, existingPath string, prefix string) string {
+	t.Helper()
+	config := filepath.Join(t.TempDir(), fmt.Sprintf("%s-ftl-project.toml", prefix))
+	var existing []byte
+	var err error
+	if existingPath == "" {
+		existing = []byte{}
+	} else {
+		existing, err = os.ReadFile(existingPath)
+		assert.NoError(t, err)
+	}
+	err = os.WriteFile(config, existing, 0600)
+	assert.NoError(t, err)
+	return config
+}
+
+// nolint
+func testAdminConfigs(
+	t *testing.T,
+	ctx context.Context,
+	envarName string,
+	admin *AdminService,
+	entries []expectedEntry,
+) {
+	t.Setenv(envarName, "eyJiYXIiOiJiYXJmb28ifQ") // bar={"bar": "barfoo"}
+
+	module := ""
+	includeValues := true
+	resp, err := admin.ConfigList(ctx, connect.NewRequest(&ftlv1.ListConfigRequest{
+		Module:        &module,
+		IncludeValues: &includeValues,
+	}))
+	assert.NoError(t, err)
+	assert.NotZero(t, resp)
+
+	configs := resp.Msg.Configs
+	assert.Equal(t, len(entries), len(configs))
+
+	for _, entry := range entries {
+		module := entry.Ref.Module.Default("")
+		ref := &ftlv1.ConfigRef{
+			Module: &module,
+			Name:   entry.Ref.Name,
+		}
+		resp, err := admin.ConfigGet(ctx, connect.NewRequest(&ftlv1.GetConfigRequest{Ref: ref}))
+		assert.NoError(t, err)
+		assert.Equal(t, entry.Value, string(resp.Msg.Value))
+	}
+}
+
+// nolint
+func testAdminSecrets(
```

Comment:
Not a comment on this PR, just an observation: looking at the amount of duplication between config and secrets for these two helper functions, it would be nice if we could get rid of some of the duplicated(ish) code by underpining some of this logic with an interface, kind of like how the `configuration` package uses generics to do dedup. Did y'all already talk about that when you were first building it? I don't have any strong thoughts here - just wondering

Reply:
Good thought! Alec touched briefly on this during our sync on AdminService and keeping these `SecretList`, `ConfigList`, etc, rpc calls separate was preferable since it seems like a Very Bad Idea™️ to conflate the two. No strong thoughts on my end either, but I can see how we might want to eventually gate secrets / add separate auth / etc

Reply:
That's a _very_ good point! SGTM